src/outpack/location_pull.py

```diff
@@ -0,0 +1,463 @@
+    )
+
+
+def _find_all_dependencies(
+    packet_ids: List[str], metadata: Dict[str, MetadataCore]
+) -> List[str]:
+    ret = set(packet_ids)
+    packets = set(packet_ids)
+    while len(packets) > 0:
+        dependency_ids = {
+            dependencies.packet
+            for packet_id in packets
+            if packet_id in metadata.keys()
+            for dependencies in (
+                []
+                if metadata.get(packet_id) is None
+                else metadata[packet_id].depends
+            )
+        }
+        packets = dependency_ids.difference(ret)
+        ret = packets.union(ret)
+
+    return sorted(ret)
+
+
+def _location_build_pull_plan_location(
+    packets: PullPlanPackets, locations: Optional[List[str]], root: OutpackRoot
+) -> List[str]:
+    location_names = location_resolve_valid(
+        locations,
+        root,
+        include_local=False,
+        include_orphan=False,
+        allow_no_locations=len(packets.fetch) == 0,
+    )
+
+    known_packets = [
+        root.index.packets_in_location(location_name)
+        for location_name in location_names
+    ]
+    missing = packets.fetch - set(itertools.chain(*known_packets))
+    if len(missing) > 0:
+        extra = missing - set(packets.requested)
+        if len(extra) > 0:
+            hint = (
+                f"{len(extra)} missing "
+                f"{pl(extra, 'packet was', 'packets were')} "
+                f"requested as "
+                f"{pl(extra, 'dependency', 'dependencies')} "
+                f"of the {pl(extra, 'one')} you asked for: "
+                f"{format_list(extra)}"
+            )
+        else:
+            # In the case where the above is used, we probably have
+            # up-to-date metadata, so we don't display this.
+            hint = "Do you need to run 'outpack_location_pull_metadata()'?"
+
+        msg = (
+            f"Failed to find {pl(missing, 'packet')} "
+            f"{format_list(missing)}\n"
+            f"Looked in {pl(location_names, 'location')} "
+            f"{format_list(location_names)}\n" + hint
+        )
+        raise Exception(msg)
+
+    return location_names
+
+
+def _location_build_pull_plan_files(
+    packet_ids: Set[str], locations: List[str], root: OutpackRoot
+) -> List[PacketFileWithLocation]:
+    metadata = root.index.all_metadata()
+    file_hashes = {
+        file.hash
+        for packet_id in packet_ids
+        for file in metadata[packet_id].files
+    }
+    n_files = len(file_hashes)
+
+    if n_files == 0:
+        return []
+
+    # Find first location within the set which contains each packet
+    # We've already checked earlier that the file is in at least 1
+    # location so we don't have to worry about that here
+    all_files = []
+    seen_hashes = set()
+    for location_name in locations:
+        location_packets = set(root.index.packets_in_location(location_name))
+        packets_in_location = location_packets & packet_ids
```

Comment:
Does this do set intersection? Did we decide on the more explicit `.intersection()` method before?

Reply:
Thanks, updated this in a couple of places and where I was using `-` for difference